src/lib/HST/Effect/InputModule.hs

```diff
@@ -0,0 +1,143 @@
+    ModuleInterface(..)
+  , ConEntry(..)
+  , TypeName
+  , ConName
+  , revertInterfaceEntry
+    -- * Effect
+  , InputModule
+    -- * Actions
+  , getInputModule
+  , getInputModuleInterface
+  , getInputModuleInterfaceByName
+    -- * Interpretations
+  , runInputModule
+  ) where
+
+import           Data.Map.Strict     ( Map )
+import qualified Data.Map.Strict     as Map
+import           Data.Maybe          ( mapMaybe )
+import           Polysemy            ( Member, Sem, interpret, makeSem )
+
+import           HST.Effect.Report   ( Report, reportFatal )
+import qualified HST.Frontend.Syntax as S
+import           HST.Util.Messages   ( Severity(Error), message )
+
+-------------------------------------------------------------------------------
+-- Aliases for Names in Module Interfaces                                    --
+-------------------------------------------------------------------------------
+-- | The name of a data type.
+type TypeName a = S.QName a
+
+-- | The name of a data constructor.
+type ConName a = S.QName a
+
+-------------------------------------------------------------------------------
+-- Module Interface Data Types                                               --
+-------------------------------------------------------------------------------
+-- | A data type for module interfaces that stores the name and the data types
+--   of a module.
+data ModuleInterface a = ModuleInterface
+  { interfaceModName   :: Maybe (S.ModuleName a)
+    -- ^ The name of the module or @Nothing@, if the module does not have a
+    --   name.
+  , interfaceDataCons  :: Map (TypeName a) [ConEntry a]
+    -- ^ A map that maps data type names to their constructors.
+  , interfaceTypeNames :: Map (ConName a) (TypeName a)
+    -- ^ A map that maps constructor names to the names of their data types.
+  }
+
+-- | An entry of the 'ModuleInterface' for a data constructor.
+data ConEntry a = ConEntry
+  { conEntryName    :: ConName a
+    -- ^ The name of the constructor.
+  , conEntryArity   :: Int
+    -- ^ The number of fields of the constructor.
+  , conEntryIsInfix :: Bool
+    -- ^ Whether the constructor should be written in infix notation or not.
+  , conEntryType    :: TypeName a
+    -- ^ The name of the data type that the constructor belongs to.
+  }
+
+-------------------------------------------------------------------------------
+-- Module Interface Utility Functions                                        --
+-------------------------------------------------------------------------------
+-- | Converts a map entry mapping a data type name to constructor entries to
+--   a list of map entries mapping the names of the given constructors to the
+--   given type name.
+revertInterfaceEntry :: (TypeName a, [ConEntry a]) -> [(ConName a, TypeName a)]
+revertInterfaceEntry (typeName, conEntry : conEntries)
+  = (conEntryName conEntry, typeName)
+  : revertInterfaceEntry (typeName, conEntries)
+revertInterfaceEntry (_, []) = []
```

Comment:
I think `invertInterfaceEntry` would be a better name. Since the type name `conEntryType` is part of the `ConEntry` a function `ConEntry a -> (ConName a, TypeName a)` would be sufficient.